controllers/rcj_soccer_referee_supervisor/referee/referee.py

```diff
@@ -1,44 +1,248 @@
+        self.progress_check = {}
+        self.penalty_area_check = {}
+        for robot in ROBOT_NAMES:
+            self.progress_check[robot] = ProgressChecker(
+                progress_check_steps, progress_check_threshold
+            )
+
+            self.penalty_area_check[robot] = PenaltyAreaChecker(
+                penalty_area_allowed_time,
+                penalty_area_reset_after,
+            )
+
+            self.robot_in_penalty_counter[robot] = 0
+
+        self.progress_check["ball"] = ProgressChecker(
+            ball_progress_check_steps, ball_progress_check_threshold
+        )
+
+        self.eventer = Eventer()
+        # Event message queue to be drawn
+        # List of Tuples of int (time) and string (message)
+        self.event_messages_to_draw: List[Tuple[int, str]] = []
+
+        self.reset_positions()
+        self.sv.update_positions()
+        self.sv.draw_team_names(self.team_name_blue, self.team_name_yellow)
+        self.sv.draw_scores(self.score_blue, self.score_yellow)
+
+    def _pack_packet(self) -> bytes:
+        """Pack data into packet.
+
+        Returns:
+            bytes: the packed packet.
+        """
+        # True/False telling whether the goal was scored
+        struct_fmt = "?"
+        data = list()
+
+        # Add Notification if the goal is scored and we are waiting for kickoff
+        # The value is True or False
+        data.append(self.ball_reset_timer > 0)
+
+        return struct.pack(struct_fmt, *data)
+
+    def _add_initial_position_noise(
+        self, translation: List[float]
+    ) -> List[float]:
+        """Return translation with noise added to x and z coordinates.
+
+        Args:
+            translation (list): x, y and z coordinates
+
+        Returns:
+            list: new x, y, and z coordinates
+        """
+        level = self.initial_position_noise
+        return [
+            translation[0] + (random.random() - 0.5) * level,
+            translation[1],
+            translation[2] + (random.random() - 0.5) * level,
+        ]
+
+    def add_event_subscriber(self, subscriber: EventHandler):
+        """Add new event subscriber.
+
+        Args:
+            subscriber (EventHandler): Instance inheriting EventHandler
+        """
+        self.eventer.subscribe(subscriber)
+
+    def add_event_message_to_queue(self, message: str):
+        """Add new message to the message queue.
+
+        Args:
+            message (str): Message to be added to the queue.
+        """
+        if len(self.event_messages_to_draw) >= MAX_EVENT_MESSAGES_IN_QUEUE:
+            self.event_messages_to_draw.pop(0)
+
+        self.event_messages_to_draw.append((self.time, message))
+
+    def process_and_draw_event_messages(self):
+        """Process and draw event messages from the queue"""
+        messages = []
+        for time, msg in self.event_messages_to_draw:
+            messages.append("{} - {}".format(time_to_string(time), msg))
```

Comment:
@Adman does it make sense to f-string this or is it just not worth it at this point?

Reply:
Makes sense, done